src/context/TodosContext.tsx

```diff
@@ -0,0 +1,108 @@
+import React, { useEffect, useReducer } from 'react';
+import { Todo } from '../interfaces/Todo';
+
+type Action = { type: 'add', todo: Todo }
+| { type: 'delete', todoId: number }
+| { type: 'change completed status', todoId: number }
+| { type: 'change title', todoId: number, newTitle: string }
+| { type: 'change completed status for all', isCompletedAll: boolean }
+| { type: 'delete completed' }
+| { type: 'set todos', todos: Todo[] | [] };
+
+export const TodosContext = React.createContext<Todo[]>([] as Todo[]);
+export const DispatchTodosContext
+= React.createContext((action: Action) => {
+  // eslint-disable-next-line no-console
+  console.log(action);
+});
+
+type Props = {
+  children: React.ReactNode;
+};
+
+function reducer(todos: Todo[], action: Action): Todo[] {
+  switch (action.type) {
+    case 'set todos':
+      return [...action.todos];
+
+    case 'add':
+      return [
+        ...todos,
+        action.todo,
+      ];
+
+    case 'change completed status': {
+      const currentTodo = todos.find(todo => todo.id === action.todoId);
+
+      if (!currentTodo) {
+        return todos;
+      }
+
+      const newTodos = [...todos];
+      const index = todos.indexOf(currentTodo);
+
+      newTodos.splice(index, 1, {
+        ...currentTodo,
+        completed: !currentTodo.completed,
+      });
+
+      return newTodos;
```

Comment:
You can make it in more elegant way using `map`

```suggestion
      return todos.map(todo => ({...todo, completed: todo.id ===  action.todoId ? !todo.completed : todo.completed}))
```